geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/iterators/DensityScan.scala

```diff
@@ -203,235 +198,110 @@ object DensityScan extends LazyLogging {
     override def hashCode(): Int = schema.hashCode()
   }
 
+  /**
+    * Gets the weight for a simple feature
+    */
+  sealed trait Weigher {
+    def weight(sf: SimpleFeature): Double
+  }
+
+  case object EqualWeight extends Weigher {
+    override def weight(sf: SimpleFeature): Double = 1d
+  }
 
   /**
-    * Gets the weight for a feature from a double attribute
+    * Gets the weight for a feature from a numeric attribute
     */
-  private def getWeightFromNumber(i: Int)(sf: SimpleFeature): Double = {
-    val d = sf.getAttribute(i).asInstanceOf[Number]
-    if (d == null) { 0.0 } else { d.doubleValue }
+  class WeightByNumber(i: Int) extends Weigher {
+    override def weight(sf: SimpleFeature): Double = {
+      val d = sf.getAttribute(i).asInstanceOf[Number]
+      if (d == null) { 0.0 } else { d.doubleValue }
+    }
   }
 
   /**
     * Tries to convert a non-double attribute into a double
     */
-  private def getWeightFromNonNumber(i: Int)(sf: SimpleFeature): Double = {
-    val d = sf.getAttribute(i)
-    if (d == null) { 0.0 } else {
-      val converted = FastConverter.convert(d, classOf[java.lang.Double])
-      if (converted == null) { 1.0 } else { converted.doubleValue() }
+  class WeightByNonNumber(i: Int) extends Weigher {
+    override def weight(sf: SimpleFeature): Double = {
+      val d = sf.getAttribute(i)
+      if (d == null) { 0.0 } else {
+        val converted = FastConverter.convert(d, classOf[java.lang.Double])
+        if (converted == null) { 1.0 } else { converted.doubleValue() }
+      }
     }
   }
 
   /**
     * Evaluates an arbitrary expression against the simple feature to return a weight
     */
-  private def getWeightFromExpression(e: Expression)(sf: SimpleFeature): Double = {
-    val d = e.evaluate(sf, classOf[java.lang.Double])
-    if (d == null) 0.0 else d
+  class WeightByExpression(e: Expression) extends Weigher {
+    override def weight(sf: SimpleFeature): Double = {
+      val d = e.evaluate(sf, classOf[java.lang.Double])
```

Comment:
Does this have any issues if the type of the attribute is an int / long / not-a-double?

Reply:
when you pass in classOf double, it tries to convert the result to a double, using the geotools Converters class.